[llvm-debuginfo-analyzer] Remove `LVScope::Children` container

jalopezg-git opened the pull request:
Remove the `LVScope::Children` container and use `llvm::concat()` instead to return a view over the types, symbols, and sub-scopes contained in a given `LVScope`.

Fixes #69160.

kuhar reviewed (changes requested):
Please submit ADT changes in a separate PR to make it easy to revert. Would be also nice to add a unit test.

CarlosAlbertoEnciso commented:
> This PR should yield further memory footprint improvements after #144399 is merged.
> 
> As suggested in #69160, removing `LVScope::Children`. Note, however, that children iteration order is now _(1)_ Types, _(2)_ Symbols, and finally _(3)_ Scopes. This may be (or not) considered important - I think it's not. ` llvm-debuginfo-analyzer` tests should be fixed though due to iteration order and `CHECK-NEXT:` FileCheck directives used in those tests.

From https://github.com/llvm/llvm-project/issues/69160#issuecomment-2987673689

The idea behind the `Children` container is to generate a logical view where its logical elements order is very close to their programming elements from the input file. That is the default printing mode.
By using the chaining iterator, we hope to have a considerable memory reduction, which is very important.
May be we can use the `LVObject::ID` to recreate the original printing order via the `LVSortFunction llvm::logicalview::getSortFunction()`.

jalopezg-git commented:
> @jalopezg-git I think the base branch for this patch should be #144744. This patch should have only the changes to the `Children` container.

That is right -- I can probably just rebase it atop `main` once https://github.com/llvm/llvm-project/pull/144744 is landed.  On this side, we are missing the element ordering (that's why some tests fail).
But, the current state should be enough to see what are the memory savings and if it pays off for the additional work :slightly_smiling_face:.  Could you run the tool w/ that big Chrome debug info file again?

jalopezg-git commented on llvm/test/tools/llvm-debuginfo-analyzer/DWARF/01-dwarf-select-logical-elements.test:
```diff
@@ -37,17 +37,16 @@
 ; ONE-NEXT: [004]           {Code} 'movl	-0x14(%rbp), %eax'
 ; ONE-NEXT: [003]           {Code} 'movl	-0x4(%rbp), %eax'
 ; ONE-NEXT: [003]     4     {TypeAlias} 'INTEGER' -> 'int'
-; ONE-NEXT: [004]     5     {Variable} 'CONSTANT' -> 'const INTEGER'
```
It was, indeed, an issue on my side.  Fixed; resolving this conversation :+1:.

CarlosAlbertoEnciso commented:
The changes that introduce the `none : unsorted output` should be in a separated patch, as it is a general feature, independent of the `children` removal and associated with `LVObject::ID`.

CarlosAlbertoEnciso reviewed (commented):
As the removal of the `Children` in somehow depends on the new `sort` options, the general documentation should include a short explanation about the `getChildren' and `getSortedChildren` to describe the concatenation order and their interaction with `sort`.

CarlosAlbertoEnciso commented on llvm/include/llvm/DebugInfo/LogicalView/Core/LVObject.h:
```diff
@@ -74,13 +74,15 @@ using LVSymbolGetFunction = bool (LVSymbol::*)() const;
 using LVScopes = SmallVector<LVScope *, 8>;
 using LVSymbols = SmallVector<LVSymbol *, 8>;
 using LVTypes = SmallVector<LVType *, 8>;
+using LVElements = SmallVector<LVElement *, 8>;
```
Can we keep the alphabetical order?
Just move the `LVElements` and `LVElementsView` before the `LVLines.`

CarlosAlbertoEnciso commented on llvm/unittests/DebugInfo/LogicalView/DWARFReaderTest.cpp:
```diff
@@ -163,13 +163,12 @@ void checkUnspecifiedParameters(LVReader *Reader) {
   EXPECT_EQ(Function->getName(), "foo_printf");
-  const LVElements *Elements = Function->getChildren();
-  ASSERT_NE(Elements, nullptr);
+  const auto Elements = Function->getChildren();
```
May be use the `LVElementsView`?

CarlosAlbertoEnciso reviewed (approved):
LGTM (with the 2 small changes).
Many thanks for the patch.

jalopezg-git committed:
[llvm-debuginfo-analyzer] Remove `LVScope::Children` container

Remove the `LVScope::Children` container and use `llvm::concat()`
instead to return a view over the types, symbols, and sub-scopes
contained in a given `LVScope`.
